src/dawn/Optimizer/PassSetBoundaryCondition.cpp

```diff
@@ -0,0 +1,332 @@
+
+    Extents fullExtent;
+    // Did we already apply a BoundaryCondition for this field?
+    // This is the first time we apply a BC to this field, we traverse all stencils that were
+    // applied before
+    std::vector<int> stencilIDsToVisit(StencilIDsVisited_);
+    if(StencilBCsApplied_.count(fieldname) == 0) {
+    } else {
+      for(int traveresedID : StencilBCsApplied_.find(fieldname)->second) {
+        stencilIDsToVisit.erase(
+            std::remove(stencilIDsToVisit.begin(), stencilIDsToVisit.end(), traveresedID),
+            stencilIDsToVisit.end());
+      }
+    }
+    for(auto& stencil : stencilInstantiation->getStencils()) {
+      for(const int& ID : stencilIDsToVisit) {
+        if(ID == stencil->getStencilID()) {
+          fullExtent.merge(
+              analyzeStencilExtents(stencil, stencilInstantiation->getAccessIDFromName(fieldname)));
+          if(StencilBCsApplied_.count(fieldname) == 0) {
+            StencilBCsApplied_.emplace(fieldname, std::vector<int>{stencil->getStencilID()});
+          } else {
+            StencilBCsApplied_.find(fieldname)->second.push_back(stencil->getStencilID());
+          }
+          break;
+        }
+      }
+    }
+    return fullExtent;
+  };
+
+  auto insertExtentsIntoMap = [](int fieldID, Extents extents, std::unordered_map<int, Extents>& map){
+      auto fieldExtentPair = map.find(fieldID);
+      if(fieldExtentPair==map.end()){
+          map.emplace(fieldID, extents);
+      }
+      else{
+          fieldExtentPair->second.merge(extents);
+      }
+  };
+
+  // Loop through all the StmtAccessPair in the stencil forward
+  for(const auto& stencilPtr : stencilInstantiation->getStencils()) {
+    Stencil& stencil = *stencilPtr;
+    DAWN_LOG(INFO) << "analyzing stencil " << stencilInstantiation->getName();
+    std::unordered_map<int, Extents> stencilDirtyFields;
+    stencilDirtyFields.clear();
+
+    for(const auto& multiStagePtr : stencil.getMultiStages()) {
+      MultiStage& multiStage = *multiStagePtr;
+      for(auto stageIt = multiStage.getStages().begin(); stageIt != multiStage.getStages().end();
+          ++stageIt) {
+        Stage& stage = (**stageIt);
+        for(const auto& domethod : stage.getDoMethods()) {
+          for(const auto& stmtAccess : domethod->getStatementAccessesPairs()) {
+            Accesses& acesses = *(stmtAccess->getAccesses());
+            const auto& allReadAccesses = acesses.getReadAccesses();
+            const auto& allWriteAccesses = acesses.getWriteAccesses();
+
+            // ReadAccesses can trigger Halo-Updates and Boundary conditions if the following
+            // criteria are fullfilled:
+            // It is a Field (ID!=-1) and we had a write before from another stencil (is in
+            // dirtyFields)
+            for(const auto& readacccess : allReadAccesses) {
+              int originalID = getOriginalID(readacccess.first);
+              if(originalID == FieldType::FT_NotOriginal)
```

Comment:
I think FT_NotOriginal is not very readable... Original is also used to refer to the original name of fields, when calling a function. 
Isnt this type of access refering to Literals? FT_Literal?

Reply:
This refers also to not-originally present fields (such as merged temporaries, cached fields). What would be a better name?